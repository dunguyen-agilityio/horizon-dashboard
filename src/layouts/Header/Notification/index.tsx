import { useRouter } from 'next/navigation';

// Components
import NotifyIcon from '@/icons/Notify';
import ListNotification from './ListNotification';
import { Popover, PopoverContent, PopoverTrigger } from '@nextui-org/react';

// Models
import { Notify } from '@/models/Notify';

interface NotificationProps {
  notifies: Notify[];
}

const Notification = ({ notifies }: NotificationProps) => {
  const { push } = useRouter();

  const handleSelect = (key: React.Key) => {
    push(key as string);
  };

  return (
    <Popover
      placement="bottom"
      showArrow={true}
      className="before:bg-gray data-[placement=bottom]:before:left-[64%] before:dark:bg-indigo-light"
    >
      <PopoverTrigger>
        <button
          className="focus-visible:outline-none aria-expanded:scale-1 aria-expanded:opacity-100"
          data-testid="notify-button"
        >
          <NotifyIcon />
        </button>
      </PopoverTrigger>
      <PopoverContent className="bg-gray dark:bg-indigo-light min-w-[320px] items-start w-full pr-[2px]">
        <ListNotification notifies={notifies} onSelect={handleSelect} />
      </PopoverContent>
    </Popover>
  );
};

export default Notification;
