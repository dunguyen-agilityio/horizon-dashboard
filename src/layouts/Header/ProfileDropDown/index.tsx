'use client';

// Libs
import { useState } from 'react';

// Components
import {
  Avatar,
  Modal,
  ModalBody,
  ModalContent,
  ModalFooter,
  ModalHeader,
  Popover,
  PopoverContent,
  PopoverTrigger,
  useDisclosure,
} from '@nextui-org/react';
import Text from '@/components/Text';
import Button from '@/components/Button';

// Types
import { TEXT_SIZE } from '@/types/text';
import { TImage } from '@/models/Image';

// Actions
import { handleSignOut } from '@/actions/auth';

interface IProfileDropdownProps {
  avatar?: TImage;
  username: string;
}

const ProfileDropDown = ({ avatar, username }: IProfileDropdownProps) => {
  const [isVisible, setIsVisible] = useState(false);

  const { isOpen, onOpen, onClose } = useDisclosure();

  const handleClose = () => {
    onClose();
  };

  const handleCloseDropDown = () => {
    setIsVisible((prev) => !prev);
  };

  const handleLogout = async () => {
    await handleSignOut();
    handleClose();
  };

  const handleCloseModal = () => {
    onOpen();
    handleCloseDropDown();
  };

  return (
    <div>
      <Popover
        placement="bottom-end"
        showArrow={true}
        isOpen={isVisible}
        onOpenChange={handleCloseDropDown}
        className="before:bg-gray before:dark:bg-indigo-light"
      >
        <PopoverTrigger>
          <button
            className="focus-visible:outline-none aria-expanded:scale-1 aria-expanded:opacity-100"
            data-testid="profile-trigger-btn"
          >
            <Avatar src={avatar?.url} alt={username} />
          </button>
        </PopoverTrigger>
        <PopoverContent
          className="bg-gray dark:bg-indigo-light"
          data-testid="profile-content"
        >
          <div className="flex flex-col gap-4 px-1 py-2 min-w-[150px]">
            <Text size={TEXT_SIZE.md} className="font-bold">
              {username}
            </Text>
            <Button
              data-testId="button-sign-out"
              className="justify-start"
              variant="flat"
              onClick={handleCloseModal}
            >
              Sign Out
            </Button>
          </div>
        </PopoverContent>
      </Popover>
      {isOpen && (
        <Modal
          isOpen
          onOpenChange={onClose}
          isDismissable={false}
          data-testid="confirm-signOut-modal"
        >
          <ModalContent className="bg-white dark:bg-content1">
            <ModalHeader className="flex flex-col gap-1">Sign Out</ModalHeader>
            <ModalBody>
              <Text>Are you sure you want to sign out?</Text>
            </ModalBody>
            <ModalFooter>
              <Button color="danger" variant="light" onPress={handleClose}>
                Close
              </Button>
              <Button
                data-testId="modal-button-sign-out"
                color="primary"
                onPress={handleLogout}
              >
                Sign Out
              </Button>
            </ModalFooter>
          </ModalContent>
        </Modal>
      )}
    </div>
  );
};

export default ProfileDropDown;
