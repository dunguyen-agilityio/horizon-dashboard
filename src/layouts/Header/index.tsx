'use client';

// Libs
import { BreadcrumbItem, Breadcrumbs } from '@nextui-org/breadcrumbs';
import { usePathname, useRouter, useSearchParams } from 'next/navigation';

// Constants
import { PRIVATE_ROUTES } from '@/constants/routes';
import { PARAMS } from '@/constants/params';

// Types
import { TEXT_SIZE } from '@/types/text';
import { User } from '@/models/User';

// Components
import Notification from './Notification';
import ProfileDropDown from './ProfileDropDown';
import { ToggleTheme, Text, InputSearch, BoxIcon } from '@/components';

// Layouts
import MenuBar from '@/layouts/MenuBar';

// Icons
import { Heart } from '@/icons';

// Utils
import { debounce } from '@/utils/debounce';

// Mocks
import { MOCK_NOTIFIES } from '@/mocks/notify';

// Hooks
import useBreadcrumbs from '@/hooks/useBreadcrumbs';

interface HeaderProps {
  isAuthenticated: boolean;
  userData: User;
}

const breadcrumbStyle =
  'dark:[&_:is(span,a)]:text-white [&_*:is(span,a)]:text-foreground';

const Header = ({ isAuthenticated, userData }: HeaderProps) => {
  const { avatar, username = '' } = userData ?? {};

  const searchParams = useSearchParams();
  const { push } = useRouter();

  const pathname = usePathname();

  const breadcrumbs = useBreadcrumbs();

  const query = searchParams.get(PARAMS.SEARCH) ?? '';

  const handleSearch = (value: string) => {
    const params = new URLSearchParams(searchParams);
    if (value) {
      params.set(PARAMS.SEARCH, value);
    } else {
      params.delete(PARAMS.SEARCH);
    }
    push(`${pathname}?${params.toString()}`);
  };

  const handleNavigateFavoritePage = () => push(PRIVATE_ROUTES.NFT_FAVORITES);

  return (
    <header className="sticky top-0 z-50 flex justify-between flex-col gap-4 sm:items-end sm:flex-row sm:gap-0 w-full xl:pl-2.5">
      <div className="flex flex-col gap-1 mr-2">
        {!!breadcrumbs.length && (
          <>
            <div className="flex justify-start w-1/2">
              <MenuBar isAuthenticated={isAuthenticated} />
            </div>
            <Breadcrumbs separator="/">
              <BreadcrumbItem className={breadcrumbStyle} isCurrent>
                Pages
              </BreadcrumbItem>
              {breadcrumbs.map(({ title, href }, index) => (
                <BreadcrumbItem
                  href={href}
                  isCurrent={index === breadcrumbs.length - 1}
                  className={breadcrumbStyle}
                  key={index}
                >
                  {title}
                </BreadcrumbItem>
              ))}
            </Breadcrumbs>
            <Text size={TEXT_SIZE['2xl']} className="mt-1">
              {breadcrumbs[0].title}
            </Text>
          </>
        )}
      </div>
      <div className="h-fit p-[10px] bg-white dark:bg-indigo rounded-[30px] flex gap-6 items-center">
        <InputSearch
          key={`search-${query}`}
          onSearch={debounce(handleSearch)}
          defaultValue={query}
        />
        <div className="absolute top-0 right-0 xl:static flex gap-2 sm:gap-6 items-center">
          {isAuthenticated && <Notification notifies={MOCK_NOTIFIES} />}
          {isAuthenticated && (
            <BoxIcon
              onClick={handleNavigateFavoritePage}
              customClass="fill-secondary dark:fill-white rounded-full justify-center items-center pt-0.5 hover:bg-default-200"
              icon={<Heart />}
            />
          )}
          <ToggleTheme />
          {isAuthenticated && userData && (
            <ProfileDropDown avatar={avatar} username={username} />
          )}
        </div>
      </div>
    </header>
  );
};

export default Header;
