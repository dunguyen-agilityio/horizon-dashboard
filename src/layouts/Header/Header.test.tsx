import { fireEvent, render } from '@testing-library/react';
import Header from '.';

const mockPush = jest.fn();

jest.mock('next/navigation', () => ({
  useRouter: jest.fn().mockImplementation(() => ({
    push: mockPush,
  })),
  useSearchParams: jest.fn().mockImplementation(() => new URLSearchParams()),
  usePathname: jest.fn().mockImplementation(() => '/'),
}));

describe('Header tests', () => {
  beforeEach(() => {
    jest.useFakeTimers();
  });

  afterEach(() => {
    jest.useRealTimers();
    jest.clearAllMocks();
  });

  it('Should match snapshot', () => {
    const { container } = render(<Header />);
    expect(container).toMatchSnapshot();
  });

  it('Should push search query correctly', async () => {
    const mockSearch = 'sea';

    const { getByTestId } = render(<Header />);

    fireEvent.change(getByTestId('input-search'), {
      target: { value: mockSearch },
    });
    jest.advanceTimersByTime(1000);
    expect(mockPush).toHaveBeenNthCalledWith(1, `/?q=${mockSearch}`);
  });

  it('Should push search query correctly', async () => {
    const mockSearch = 'sea';

    const { getByTestId } = render(<Header />);

    fireEvent.change(getByTestId('input-search'), {
      target: { value: mockSearch },
    });

    fireEvent.change(getByTestId('input-search'), {
      target: { value: '' },
    });

    jest.advanceTimersByTime(1000);
    expect(mockPush).toHaveBeenNthCalledWith(1, '/?');
  });
});
