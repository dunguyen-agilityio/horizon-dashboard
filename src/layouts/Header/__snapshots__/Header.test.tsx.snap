// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header tests Should match snapshot 1`] = `
<div>
  <header
    class="pt-4 md:pt-[50px] sticky top-0 z-30 bg-gray dark:bg-indigo-dark flex justify-between flex-col gap-4 sm:items-end sm:flex-row sm:gap-0 w-full xl:pl-2.5"
  >
    <div
      class="flex flex-col gap-1 mr-2"
    />
    <div
      class="h-fit p-[10px] bg-white dark:bg-indigo rounded-[30px] flex gap-6 items-center"
    >
      <div
        class="group flex flex-col data-[hidden=true]:hidden w-full group relative justify-end data-[has-label=true]:mt-[calc(theme(fontSize.small)_+_10px)]"
        data-filled="true"
        data-filled-within="true"
        data-slot="base"
      >
        <div
          class="relative w-full inline-flex tap-highlight-transparent flex-row items-center shadow-sm px-3 gap-3 data-[hover=true]:bg-default-200 h-10 min-h-10 transition-background motion-reduce:transition-none !duration-150 outline-none group-data-[focus-visible=true]:z-10 group-data-[focus-visible=true]:ring-2 group-data-[focus-visible=true]:ring-focus group-data-[focus-visible=true]:ring-offset-2 group-data-[focus-visible=true]:ring-offset-background md:w-96 xl:w-fit bg-gray group-data-[focus=true]:bg-gray dark:bg-indigo-dark dark:group-data-[focus=true]:bg-indigo-dark rounded-[30px]"
          data-slot="input-wrapper"
          data-testid="search-wrapper"
          style="cursor: text;"
        >
          <div
            class="w-full items-center h-full box-border flex gap-[11px] pl-[8px]"
            data-slot="inner-wrapper"
          >
            <span>
              <svg
                class="[&>line]:stroke-primary [&>line]:dark:stroke-white [&>circle]:stroke-primary [&>circle]:dark:stroke-white"
                fill="none"
                height="12"
                viewBox="0 0 11 12"
                width="11"
                xmlns="http://www.w3.org/2000/svg"
              >
                <circle
                  cx="5"
                  cy="5"
                  r="4.3"
                  stroke="#2B3674"
                  stroke-width="1.4"
                />
                <line
                  stroke="#2B3674"
                  stroke-linecap="round"
                  stroke-width="1.4"
                  x1="10.0101"
                  x2="8"
                  y1="11"
                  y2="8.98995"
                />
              </svg>
            </span>
            <input
              aria-label="Search"
              class="w-full font-normal bg-transparent !outline-none placeholder:text-foreground-500 focus-visible:outline-none data-[has-start-content=true]:ps-1.5 data-[has-end-content=true]:pe-1.5 file:cursor-pointer file:bg-transparent file:border-0 autofill:bg-transparent bg-clip-text text-small group-data-[has-value=true]:text-default-foreground placeholder-purple-50 dark:placeholder-white text-primary dark:text-white"
              data-has-start-content="true"
              data-slot="input"
              data-testid="input-search"
              id="react-aria-:r0:"
              placeholder="Search"
              type="search"
              value=""
            />
          </div>
        </div>
      </div>
      <div
        class="absolute top-4 md:top-14 right-0 xl:static flex gap-2 sm:gap-6 items-center"
      >
        <button
          aria-expanded="false"
          aria-label="notifications-button"
          class="z-10 subpixel-antialiased focus-visible:outline-none aria-expanded:scale-1 aria-expanded:opacity-100"
          data-slot="trigger"
          data-testid="notify-button"
          type="button"
        >
          <i
            class="w-6 h-6 flex justify-center items-center hover:cursor-pointer fill-secondary dark:fill-white"
          >
            <svg
              height="20"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M15.29 15.29 14 14V9c0-3.07-1.64-5.64-4.5-6.32V2C9.5 1.17 8.83.5 8 .5S6.5 1.17 6.5 2v.68C3.63 3.36 2 5.92 2 9v5L.71 15.29c-.63.63-.19 1.71.7 1.71h13.17c.9 0 1.34-1.08.71-1.71ZM12 15H4V9c0-2.48 1.51-4.5 4-4.5s4 2.02 4 4.5v6Zm-4 5c1.1 0 2-.9 2-2H6a2 2 0 0 0 2 2Z"
              />
            </svg>
          </i>
        </button>
        <i
          class="w-6 h-6 flex hover:cursor-pointer fill-secondary dark:fill-white rounded-full justify-center items-center pt-0.5 hover:bg-default-200"
        >
          <svg
            height="15"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M13.944 1.4C11.959-.008 9.509.65 8.186 2.261 6.863.65 4.413-.016 2.43 1.401 1.376 2.15.715 3.42.669 4.757.566 7.795 3.15 10.23 7.097 13.963l.075.07c.572.54 1.451.54 2.022-.007l.083-.079c3.946-3.725 6.524-6.16 6.427-9.197-.046-1.33-.707-2.598-1.76-3.35ZM8.26 12.797l-.075.078-.075-.078C4.533 9.423 2.173 7.192 2.173 4.93c0-1.565 1.128-2.74 2.63-2.74 1.158 0 2.286.776 2.684 1.848h1.406c.39-1.072 1.518-1.847 2.675-1.847 1.504 0 2.631 1.174 2.631 2.74 0 2.261-2.36 4.492-5.938 7.866Z"
            />
          </svg>
        </i>
        <button
          aria-label="theme-button"
          class="z-0 group relative box-border appearance-none select-none whitespace-nowrap font-normal subpixel-antialiased overflow-hidden tap-highlight-transparent data-[pressed=true]:scale-[0.97] outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 text-small gap-2 [&>svg]:max-w-[theme(spacing.8)] transition-transform-colors-opacity motion-reduce:transition-none text-default-foreground data-[hover=true]:opacity-hover min-w-fit h-fit rounded-full flex justify-center items-center p-2 bg-transparent"
          type="button"
        >
          <svg
            class="[&_path]:fill-secondary dark:[&_path]:fill-white [&_rect]:fill-secondary dark:[&_rect]:fill-white"
            data-testid="moon-icon"
            fill="none"
            height="18"
            viewBox="0 0 18 18"
            width="18"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g
              clip-path="url(#clip0_101_12742)"
            >
              <path
                d="M9.95703 18C12.733 18 15.2684 16.737 16.9481 14.6675C17.1966 14.3613 16.9256 13.9141 16.5416 13.9872C12.1751 14.8188 8.16522 11.4709 8.16522 7.06303C8.16522 4.52398 9.52444 2.18914 11.7335 0.931992C12.074 0.738211 11.9884 0.221941 11.6015 0.150469C11.059 0.0504468 10.5086 8.21369e-05 9.95703 0C4.98914 0 0.957031 4.02578 0.957031 9C0.957031 13.9679 4.98281 18 9.95703 18Z"
              />
            </g>
            <defs>
              <clippath
                id="clip0_101_12742"
              >
                <rect
                  height="18"
                  width="18"
                />
              </clippath>
            </defs>
          </svg>
        </button>
      </div>
    </div>
  </header>
</div>
`;
